addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,434 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	kapppkgiv1alpha1 "github.com/vmware-tanzu/carvel-kapp-controller/pkg/apis/packaging/v1alpha1"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/types"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+	"github.com/vmware-tanzu/tanzu-framework/addons/predicates"
+	runtanzuv1alpha3 "github.com/vmware-tanzu/tanzu-framework/apis/run/v1alpha3"
+	tkrconstants "github.com/vmware-tanzu/tanzu-framework/pkg/v1/tkr/pkg/constants"
+)
+
+type ClusterRole int
+
+const (
+	clusterRoleManagement ClusterRole = iota
+	clusterRoleWorkload
+)
+
+// PackageInstallStatusReconciler reconciles a PackageInstallStatus object
+// PackageInstallStatusReconciler is responsible for removing remote cluster from watches when
+// the cluster is being deleted.
+type PackageInstallStatusReconciler struct {
+	Ctx        context.Context
+	Log        logr.Logger
+	Client     client.Client
+	Scheme     *runtime.Scheme
+	controller controller.Controller
+	Tracker    *util.RemoteObjectTracker
+	liveClient client.Client
+}
+
+//+kubebuilder:rbac:groups=addons.tanzu.vmware.com,resources=packageinstallstatuses,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=addons.tanzu.vmware.com,resources=packageinstallstatuses/status,verbs=get;update;patch
+
+// Reconcile performs the reconciliation action for the controller
+func (r *PackageInstallStatusReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.Log.WithValues(constants.ClusterNamespaceLogKey, req.Namespace, constants.ClusterNameLogKey, req.Name)
+	log.V(4).Info("Reconciling")
+
+	var (
+		c           client.Client
+		clusterRole ClusterRole
+	)
+
+	// get cluster object
+	cluster := &clusterapiv1beta1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Cluster not found")
+			r.Tracker.DeleteAccessor(req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "unable to fetch cluster")
+		return ctrl.Result{}, err
+	}
+
+	// if cluster is marked for deletion, then no reconciliation is needed; just remove the cluster object from the cluster accessor map
+	if !cluster.GetDeletionTimestamp().IsZero() {
+		r.Tracker.DeleteAccessor(req.NamespacedName)
+		return ctrl.Result{}, nil
+	}
+
+	clusterLabels := cluster.GetLabels()
+
+	// make sure the TKR object exists
+	tkrName := clusterLabels[constants.TKRLabelClassyClusters]
+	tkr, err := util.GetTKRByName(ctx, r.Client, tkrName)
```

Comment:
question: why do we use `r.Client` instead of `r.liveClient` here ?

Reply:
liveClient  is needed when it is required that the client read directly from the server and not to use object caches. For Package and PackaheInstalls which get available/installed in the workload cluster after the remote kapp controller package is ready, it's not a good idea to use a client which can cache for read. For TKR, it doesn't really matter which client to use